cytominer_eval/operations/util.py

```diff
@@ -131,3 +133,87 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
         len(entries.unique()) == 1
     ), "grit is calculated for each perturbation independently"
     return str(list(entries)[0])
+
+
+class DistributionEstimator:
+    def __init__(self, arr):
+        self.sigma = EmpiricalCovariance().fit(arr)
+        
+    def mahalanobis(self, X):
+        return(self.sigma.mahalanobis(X))
+
+
+def calculate_mahalanobis(
+    pert_df: pd.DataFrame, control_df: pd.DataFrame
+    ) -> pd.Series:
+    """
+    Usage: Designed to be called within a pandas.DataFrame().groupby().apply()
+    """
+    assert len(control_df) > 1, "Error! No control perturbations found."
+        
+    # Get dispersion and center estimators for the 
+    control_estimators = DistributionEstimator(control_df)
+    
+    # Distance between mean of perturbation and control
+    maha = control_estimators.mahalanobis(
+                np.array(np.mean(pert_df, 0)).reshape(1, -1))[0]
+    return maha
+
+
+def default_mp_value_parameters():    
+    params = {"rescale_pca": True,
+              "nb_permutations": 100}
+    return(params)
+    
+    
+def calculate_mp_value(
+    pert_df: pd.DataFrame, control_df: pd.DataFrame,
+    params = {}
+    ) -> pd.Series:
+    """
+    Usage: Designed to be called within a pandas.DataFrame().groupby().apply()
+    """
+    assert len(control_df) > 1, "Error! No control perturbations found."
+    
+    # Assign parameters
+    p = default_mp_value_parameters()
+    assert (
+        all([x in p.keys() for x in params.keys()])
+    ), "Unknown parameters provided." 
+    for (k,v) in params.items():
+        p[k] = v
+    
+    merge_df = pert_df.append(control_df)
+    
+    # We reduce the dimensionality with PCA
+    # so that 90% of the variance is conserved
+    pca = PCA(n_components = 0.9, svd_solver = 'full')
+    pca_array = pca.fit_transform(merge_df)
+    # We scale columns by the variance explained
+    if p["rescale_pca"]:
+        pca_array = pca_array * pca.explained_variance_ratio_
+    # NB: this seems useless, as the point of using the Mahalanobis
```

Comment:
why are we adding it then? 😆

Reply:
Why did they do it in the article in the first place? 😅 
It doesn't change the results at all in the current implementation (and a single matrix multiplication doesn't hurt much), so I left the options open for now...

Reply:
cool, sounds good. What dose `NB` mean? We should delete it if it is not adding anything.

Reply:
It stands for *nota bene* and is a habit I have to signify that these comments are giving contextual information rather than directly describing the following lines of code. You're right to point out this is not conventional so I'll remove them.